### PR TITLE
chore: add contributing guide

### PR DESCRIPTION
**Pull request by aloknerurkar**
### Checklist

- [ ] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md)
- [ ] My change requires a documentation update and I have done it
- [ ] I have added tests to cover my changes.
- [ ] I have filled out the description and linked the related issues

### Description
Adding a contribution guide for the open source community. Adding it here would make it available on the Github page.

- - -
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/3222)

### CODE REVIEW

**Review by notanatol — APPROVED** [review]

**Comment by vladopajic** [comment]
Documents `CODING.md` and `CODINGSTYLE.md` revolve around same topic. Wouldn't it be better if those are merged to single document?

**Comment by aloknerurkar** [comment]
> Documents `CODING.md` and `CODINGSTYLE.md` revolve around same topic. Wouldn't it be better if those are merged to single document?

Not sure if we are looking at the same docs.

CODING doc provides information about writing code.
CODINGSTYLE doc provides information about go specific details.
CONTRIBUTING doc provides information on how to contribute to the repo. The things mentioned here are directed towards the information required for certain things.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+- Code refactorings
+
+  In the case where you manage to spend enough time in the codebase to figure out that some stuff can be refactored for idiomacy. This is not strictly discouraged. However, such changes are usually very subjective. So again the right way to go about these things is to clear them up in an issue before attempting any changes. As we mentioned before, bee is a largely stable codebase and we prefer stability over most other things. There are some packages in the codebase on which a lot of other packages depend on. Also there is some historical context for some of the decisions. So it is important to involve the team beforehand in understanding that. Some of these packages also have equivalent javascript implementations. So they also need to be consistent across codebases. It would be difficult to point out all the packages here, but some examples would be:
+
+  - `swarm` package which includes the core definitions of types that we use all around the codebase.
+
+  - Core DISC abstractions
```
Comment (aloknerurkar) [inline CONTRIBUTING.md:54]: DISC doesnt seem right here

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    Protocols like `PushSync`, `PullSync`, `hive`, `chainsync`, `retrieval`.
+
+  - Core network abstractions
+
+    `p2p`, `topology`, `accouting`.
+
+  - Core storage abstractions
+
+    `storage`, `localstore`, `postage`.
+
```
Comment (aloknerurkar) [inline CONTRIBUTING.md:65]: Add another point about adding new libraries to the codebase

**Review by mrekucci — COMMENTED** [review]
It would also be good to include a paragraph on naming stating that naming is difficult, but we prefer the consistency of naming throughout the code base. It's similar to the `gofmt` tool, the formatting isn't to everyone's liking, but it is consistent.

> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/3222#-:-NAJ48_6EO6Z4yZL1m4c:bm8llul)* status: all files reviewed (commit messages unreviewed), 2 unresolved discussions (waiting on @agazso, @istae, @janos, @swarmHaseeb, and @umerm-work)

___
*[`CONTRIBUTING.md` line 15 at r3](https://reviewable.io/reviews/ethersphere/bee/3222#-NAIw4VL5RN8T3roG9wr:-NAIw4VL5RN8T3roG9ws:b5bsxh6) ([raw file](https://github.com/ethersphere/bee/blob/f74719fe618ee7deb3acf9d0adba0ef69777e7a0/CONTRIBUTING.md#L15)):*
> ```Markdown
> ## Issues
> 
> We use github issues for planning at present. The first interaction with the codebase should be creating an issue on github. There are different issue templates already defined in the repository. Please use the appropriate template and fill the required information.
> ```

`github` -> Github.
___
*[`CONTRIBUTING.md` line 34 at r3](https://reviewable.io/reviews/ethersphere/bee/3222#-NAIwi9w3eUrNJAvoJzi:-NAIwi9w3eUrNJAvoJzj:bmzc8hz) ([raw file](https://github.com/ethersphere/bee/blob/f74719fe618ee7deb3acf9d0adba0ef69777e7a0/CONTRIBUTING.md#L34)):*
> ```Markdown
> - Good first issues
> 
>   These are work items that are low on priority for the bee team at the moment, but are definitely a nice to have. You can find the `good first issue` label in our issues. We will keep adding these from time to time. So if you don't know what to work on, it is highly recommended to pick something with this label to start.
> ```

I'd rephrase the following "...are low on priority for the bee team at the moment, but are definitely a nice to have." To something like: "...that are good if you are contributing to the codebase for the first time".
___
*[`CONTRIBUTING.md` line 36 at r4](https://reviewable.io/reviews/ethersphere/bee/3222#-NAIy-0619DXMVClodxr:-NAIy-0619DXMVClodxs:bckwaze) ([raw file](https://github.com/ethersphere/bee/blob/512f0eab690ae2a53871b8972298eec1577e431e/CONTRIBUTING.md#L36)):*
> ```Markdown
>   These are work items that are low on priority for the bee team at the moment, but are definitely a nice to have. You can find the `good first issue` label in our issues. We will keep adding these from time to time. So if you don't know what to work on, it is highly recommended to pick something with this label to start.
> 
> - Performance optimizations
> ```

I would add here that code readability takes precedence over performance optimization, which increases code complexity.
___
*[`CONTRIBUTING.md` line 38 at r4](https://reviewable.io/reviews/ethersphere/bee/3222#-NAIxUIiBZTEb6J6B8ul:-NAIxUIiBZTEb6J6B8um:b-1ycm3g) ([raw file](https://github.com/ethersphere/bee/blob/512f0eab690ae2a53871b8972298eec1577e431e/CONTRIBUTING.md#L38)):*
> ```Markdown
> - Performance optimizations
> 
>   The best way to propose any optimizations would be to provide the relevant data to describe the problem and then also the same data after the optimizations are done. Keep in mind, bee nodes work in a distributed system, so changes that would seem good locally may not hold in some cases. The bee client in debug mode can show you metrics as well as pprof information. This can be used to demonstrate the optimizations.
> ```

`bee` -> 'Bee'
___
*[`CONTRIBUTING.md` line 69 at r4](https://reviewable.io/reviews/ethersphere/bee/3222#-NAJ4-Vt2sTZuA36ojhM:-NAJ4-Vt2sTZuA36ojhN:b8bprdk) ([raw file](https://github.com/ethersphere/bee/blob/512f0eab690ae2a53871b8972298eec1577e431e/CONTRIBUTING.md#L69)):*
> ```Markdown
> ### Lifecycle of tasks
> 
> Once you have achieved the said consensus on the tasks you can start working on them. The basic git flow is expected:
> ```

`git` -> `Git`.
___
*[`.github/CONTRIBUTING.md` line 9 at r1](https://reviewable.io/reviews/ethersphere/bee/3222#-NAIvcZO2PJXe0LOVErn:-NAIvcZO2PJXe0LOVEro:b5isa6d) ([raw file](https://github.com/ethersphere/bee/blob/a31757c6b85f491853c23d2497b6899e3da73792/.github/CONTRIBUTING.md#L9)):*
> ```Markdown
> ## Getting started
> 
> To get an overview of the project go through the [README](../README.md) file. Bee client is the go implementation of the Swarm decentralized storage layer as described in [The Book of Swarm](). It is important to understand the concepts in this book before you work with the codebase.
> ```

I'd suggest using `Golang` instead of `go` since it can be misleading.

**Review by janos — APPROVED** [review]

**Review by mrekucci — APPROVED** [review]

**Review by umerm-work — APPROVED** [review]
